[mkdocs] docs: add listen tx flow tutorial (#916)
NemProject/nem: mkdocs/pages/en/devbook/websockets/listen-transaction-flow.md

## Code Explanation

### Setting Up the Monitored Address and Signer

{{ tutorial.code_snippet_tagged('step-1') }}

Each transaction channel is scoped to a specific address.
The channels send a notification whenever this address is involved in a transaction, as sender or recipient.
The `MONITOR_ADDRESS` environment variable sets the address to watch.
The WebSocket API expects it uppercase and without hyphens.

To trigger notifications, this tutorial sends a transfer transaction to the monitored address.
The sender's private key is read from `SIGNER_PRIVATE_KEY`.

If any of these environment variables is not provided, the tutorial provides default values.

### Connecting to the WebSocket

{{ tutorial.code_snippet_tagged('step-2') }}

The code opens a SockJS connection to the `/w/messages` endpoint on `WS_URL` and starts a STOMP session over it.

### Registering the Account

{{ tutorial.code_snippet_tagged('step-3') }}

To receive notifications on an account's transaction channels, the address must first be **registered** with the node.

The code registers the address by sending a <req:w&#47;api&#47;account&#47;get> request, which both registers it and

> replies with the account's current state on the <ws:account&#47;{address}> channel.

Reviewer: ```suggestion
triggers a reply containing the account's current state on the <ws:account&#47;{address}> channel.
```

To catch that reply, the code subscribes to the channel using `id-0`, then waits.

> Once the reply arrives, the registration is active and the temporary subscription is dropped.

Reviewer: ```suggestion
To verify that the subscription is successful, the code first temporarily subscribes to the
<ws:account&#47;{address}> channel, sends the registration request, and waits for that reply.

Once the reply arrives, the registration is active and the temporary subscription to the
general account messages channel is dropped, using the same `id-0` used during registration.
```

### Subscribing to the Channels

{{ tutorial.code_snippet_tagged('step-4') }}

> With the account registered, the code subscribes to two address-scoped channels:

Reviewer: ```suggestion
With the account's address registered, the code subscribes to two address-scoped channels:
```

So the connection between the account and the channels below is extra clear.

* <ws:unconfirmed&#47;{address}>: Notifies when a transaction involving the address enters the <unconfirmed pool:>,
    waiting to be included in a block.
* <ws:transactions&#47;{address}>: Notifies when a transaction involving the address is included in a <block:>.

The subscriptions use IDs `id-1` and `id-2`, which identify them when the code unsubscribes after confirmation.

### Building and Signing a Transfer Transaction

{{ tutorial.code_snippet_tagged('step-5') }}

> This tutorial builds a minimal [Transfer Transaction](../transactions/transfer-xem.md) to the monitored address, with a

Reviewer: I would make this link point to the Textbook, and then, below, instead of "as usual", I would say "following the same procedure as in the Transfer Transaction Tutorial", and include the tutorial link there.

zero amount, no mosaics, and no message.
A transfer is used for simplicity, but any transaction type triggers the same WebSocket notifications.

The transaction is built as usual: fetching the network time, creating the transaction, and signing it.
The hash is computed locally so it can be matched against incoming WebSocket messages later.

### Announcing and Waiting for Confirmation

{{ tutorial.code_snippet_tagged('step-6') }}

The code announces the transaction to the <post:/transaction/announce> endpoint and checks the result.
If the node rejected it, the code prints the rejection reason and stops.

Otherwise, the code waits for confirmation, printing each message from the two subscribed channels.
Each message follows the [TransactionMetaDataPair](../reference/rest/nem.md#model/TransactionMetaDataPair) schema, whose
`meta.hash.data` field holds the transaction hash.

> When a message from the `/transactions/{address}` channel arrives whose hash matches the announced transaction, the

Reviewer: ```suggestion
When a message from the <ws:transactions&#47;{address}> channel arrives whose hash matches the announced transaction, the
```

Same in another place below.

Each transaction channel is scoped to a specific address.
The channels send a notification whenever this address is involved in a transaction, as sender or recipient.
The `MONITOR_ADDRESS` environment variable sets the address to watch.
The WebSocket API expects it uppercase and without hyphens.

To trigger notifications, this tutorial sends a transfer transaction to the monitored address.
The sender's private key is read from `SIGNER_PRIVATE_KEY`.

If any of these environment variables is not provided, the tutorial provides default values.

### Connecting to the WebSocket

{{ tutorial.code_snippet_tagged('step-2') }}

The code opens a SockJS connection to the `/w/messages` endpoint on `WS_URL` and starts a <STOMP session:>
over it.

### Registering the Account

{{ tutorial.code_snippet_tagged('step-3') }}

To receive notifications on an account's transaction channels, the address must first be **registered** with the node.

The code registers the address by sending a <req:w&#47;api&#47;account&#47;get> request, which both registers it and
triggers a reply containing the account's current state on the <ws:account&#47;{address}> channel.

> The code first temporarily subscribes to the <ws:account&#47;{address}> channel, sends the registration request, and

Reviewer: This is a convoluted process (it requires a temporary subscription), that's why I suggested to explain why this is needed ("To verify that the subscription is successful...")

I would still like that we explained why this weird process is needed.

Author: I see where the confusion comes from. What was missing was a clearer explanation of **why** we subscribe to the channel and wait for the notification.

The way it was presented was also hard to follow since I introduced the request first, then the subscription, and later showed how the request was used.

I’ve updated the text to follow the code flow step by step. It should be clearer now.

Reviewer: I like this version much better, yup!
